File: src/DesignPatterns/Behavioural/DIY/Iterator/Iterator.java
package DesignPatterns.Behavioural.DIY.Iterator;

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

interface IIterator{
    boolean hasNext();
    String next();
    
}

class ListIterator implements IIterator{
    private List<String> items;
    private int index;
    
    public ListIterator(List<String> items){
        this.items = items;
        this.index = 0;
    }

    @Override
    public boolean hasNext() {
        return (index < items.size());
    }

    @Override
    public String next() {
        if(hasNext()){
            return items.get(index++);
        }else{
            throw new NoSuchElementException("List is empty.");
        }
    }
    
}

class Checklist{
    private List<String> items;

    public Checklist() {
        this.items = new ArrayList<>();
    }

    public void addItem(String item){
        items.add(item);
    }
    public void removeItem(String item){
        items.remove(item);
    }
    public ListIterator getIterator(){
        return new ListIterator(items);
    }
}

public class Iterator {
    public static void main(String[] args) {

        Checklist checklist = new Checklist();

        checklist.addItem("Prepare breakfast");
        checklist.addItem("Wash the dishes.");
        checklist.addItem("Do laundry.");

        ListIterator iterator = checklist.getIterator();

        System.out.println("Chores for the day: ");
        while(iterator.hasNext()){
            String item = iterator.next();
            System.out.println("+  " + item);
        }
        
    }
}

File: bin/DIY/Iterator/Iterator.ts
// Iterator.ts

interface IIterator {
    hasNext(): boolean;
    next(): string;
}

class ListIterator implements IIterator {
    private items: string[];
    private index: number;

    constructor(items: string[]) {
        this.items = items;
        this.index = 0;
    }

    hasNext(): boolean {
        return this.index < this.items.length;
    }

    next(): string {
        if (this.hasNext()) {
            return this.items[this.index++];
        } else {
            throw new Error("List is empty.");
        }
    }
}

class Checklist {
    private items: string[];

    constructor() {
        this.items = [];
    }

    addItem(item: string): void {
        this.items.push(item);
    }

    removeItem(item: string): void {
        const index = this.items.indexOf(item);
        if (index !== -1) {
            this.items.splice(index, 1);
        }
    }

    getIterator(): ListIterator {
        return new ListIterator(this.items);
    }
}

// main
(function main() {
    const checklist = new Checklist();

    checklist.addItem("Prepare breakfast");
    checklist.addItem("Wash the dishes.");
    checklist.addItem("Do laundry.");

    const iterator = checklist.getIterator();

    console.log("Chores for the day:");
    while (iterator.hasNext()) {
        const item = iterator.next();
        console.log("+  " + item);
    }
})();
